aptos-move/framework/aptos-framework/sources/code.move

```diff
@@ -200,6 +212,54 @@ module aptos_framework::code {
         }
     }
 
+    /// Check that the upgrade policies of all packages are equal or higher quality than this package.
+    fun check_dependencies(publish_address: address, pack: &PackageMetadata) acquires PackageRegistry {
+        let deps = &pack.deps;
+        let i = 0;
+        let n = vector::length(deps);
+        while (i < n) {
+            let dep = vector::borrow(deps, i);
+            assert!(exists<PackageRegistry>(dep.account), error::not_found(EPACKAGE_DEP_MISSING));
+            if (is_policy_exempted_address(dep.account)) {
+                i = i + 1;
+                continue
+            };
+            let registry = borrow_global<PackageRegistry>(dep.account);
+            let j = 0;
+            let m = vector::length(&registry.packages);
+            let found = false;
+            while (j < m) {
+                let dep_pack = vector::borrow(&registry.packages, j);
+                if (dep_pack.name == dep.package_name) {
+                    // Found the package in the registry, now check policy
+                    assert!(
+                        dep_pack.upgrade_policy.policy >= pack.upgrade_policy.policy,
+                        error::invalid_argument(EDEP_WEAKER_POLICY)
+                    );
+                    if (dep_pack.upgrade_policy == upgrade_policy_arbitrary()) {
+                        assert!(
+                            dep.account == publish_address,
+                            error::invalid_argument(EDEP_ARBITRARY_NOT_SAME_ADDRESS)
+                        )
+                    };
+                    found = true;
+                    break
+                };
+                j = j + 1;
+            };
+            assert!(found, error::not_found(EPACKAGE_DEP_MISSING));
+            i = i + 1;
+        };
+    }
+
+    /// Core addresses which are exempted from the check that their policy matches the referring package. Without
+    /// this exemption, it would not be possible to define an immutable package based on the core system, which
+    /// requires to be upgradable for maintenance and evolution, and is configured to be `compatible`.
+    fun is_policy_exempted_address(addr: address): bool {
+        // TODO: put this perhaps on-chain
+        addr == @1 || addr == @2
```

Comment:
This seems a bit hacky. @2 is also not the right address. Token is deployed at @3. I think this should check against all addresses that Aptos governance owns

Reply:
I know. Is this already publicly available on chain?

FWIW, someone pointed out on an issue that we count `0x1, 0x2, ..., 0x9, 0x10` instead of `0xa`. That was in some genesis code. Should we fix that?

Reply:
I can send a quick fix. You can use this function here https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-framework/sources/system_addresses.move#L33

Reply:
Done